File: svc/src/main/java/com/yada/ssp/manager/svc/config/SecurityConfig.java
package com.yada.ssp.manager.svc.config;

import org.springframework.context.annotation.Bean;
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.config.core.GrantedAuthorityDefaults;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;

import java.util.List;
import java.util.stream.Collectors;

@EnableWebSecurity
@EnableGlobalMethodSecurity(securedEnabled=true)
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http
                .authorizeRequests(registry ->  registry.anyRequest().authenticated())
                .oauth2ResourceServer()
                .jwt()
                .jwtAuthenticationConverter(it -> {
                    JwtAuthenticationConverter jwtConverter = new JwtAuthenticationConverter();
                    jwtConverter.setJwtGrantedAuthoritiesConverter(jwt ->
                            ((List<String>) jwt.getClaimAsMap("realm_access").get("roles"))
                                    .stream()
                                    .map(SimpleGrantedAuthority::new)
                                    .collect(Collectors.toList())
                    );
                    return jwtConverter.convert(it);
                });
    }

    @Bean
    GrantedAuthorityDefaults grantedAuthorityDefaults() {
        return new GrantedAuthorityDefaults(""); // Remove the ROLE_ prefix
    }
}

File: webSvc/src/main/kotlin/com/yada/ssp/manager/web/config/SecurityConfig.kt
package com.yada.ssp.manager.web.config

import com.yada.ssp.manager.web.security.IPwdDigestService
import com.yada.ssp.manager.web.security.IPwdStrengthService
import com.yada.ssp.manager.web.security.PwdDigestService
import com.yada.ssp.manager.web.security.PwdStrengthService
import org.jasypt.encryption.StringEncryptor
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter
import org.springframework.security.oauth2.server.resource.authentication.ReactiveJwtAuthenticationConverterAdapter
import org.springframework.security.web.server.SecurityWebFilterChain

@Configuration
@EnableWebFluxSecurity
@EnableConfigurationProperties(SecurityConfigProperties::class)
class SecurityConfig(
    private val config: SecurityConfigProperties,
    private val stringEncryptor: StringEncryptor
) {

    @Bean
    fun pwdDigestService(): IPwdDigestService = PwdDigestService(config.defaultPwd, stringEncryptor)

    @Bean
    fun pwdStrengthService(): IPwdStrengthService = PwdStrengthService(config.pwdStrength)

    @Bean
    fun configure(http: ServerHttpSecurity): SecurityWebFilterChain {
        http
            .authorizeExchange { exchanges ->
                config.roles.keys.forEach {
                    exchanges.pathMatchers(it).hasAnyRole(*config.roles[it] ?: arrayOf())
                }
                exchanges.anyExchange().authenticated()
            }
            .oauth2ResourceServer()
            .jwt().jwtAuthenticationConverter {
                val jwtConverter = JwtAuthenticationConverter()
                jwtConverter.setJwtGrantedAuthoritiesConverter { jwt: Jwt ->
                    (jwt.getClaimAsMap("realm_access")["roles"] as List<*>)
                        .map { role -> "ROLE_$role" }
                        .map { role -> SimpleGrantedAuthority(role) }
                }
                ReactiveJwtAuthenticationConverterAdapter(jwtConverter).convert(it)
            }

        return http.build()
    }
}
